File: FA.py
import pandas as pd
from graphviz import Digraph

EPS = "ε"
EOF = "$"


class FA:
    def __init__(self, S: list, sigma: list, delta: list, s0, A: list):
        self.S = S
        self.sigma = sigma
        self.delta = delta
        self.s0 = s0
        self.A = A
        self.G = {i: {} for i in S}
        for u, w, v in delta:
            self.G[u].setdefault(w, []).append(v)

    @classmethod
    def load(cls, filename: str):
        with open(filename, "r") as f:
            S, sigma, delta, s0, A = f.read().replace("eps", EPS).split("\n\n")
            S = S.split(" ")
            sigma = sigma.split(" ")
            delta = [tuple(item.split(" ")) for item in delta.split("\n")]
            A = A.split(" ")
            return cls(S, sigma, delta, s0, A)

    def __str__(self):
        return f"(S={self.S}\nsigma={self.sigma}\ndelta={self.delta}\ns0={self.s0}\nA={self.A})"

    def move(self, s: str, w: str):
        return self.G[s][w] if w in self.G[s] else []

    def dot(self, filename: str, view: bool = False):
        dot = Digraph(filename)
        for u in self.S:
            dot.node(u, u, shape="circle")
        for u in self.A:
            dot.node(u, u, shape="doublecircle")
        for u, w, v in self.delta:
            dot.edge(u, v, label=w)
        dot.render(f"./output/{filename}.gv", view=view)


class NFA(FA):

    def dfs_eps_closure(self):
        """离线计算 eps_closure"""
        E = {s: {} for s in self.S}

        def dfs(x):
            if E[x]:
                return
            E[x] = {x}
            for u, w, v in self.delta:
                if x == u and v != u and w == EPS:
                    dfs(v)
                    E[u] = E[u] | E[v]

        for s in self.S:
            dfs(s)
        print(E)

    def eps_closure(self):
        """离线计算 eps_closure"""
        E = {s: {s} for s in self.S}
        worklist = self.S.copy()
        iter_cnt = 0
        while worklist:
            x = worklist.pop(0)
            t = {x}
            for u, w, v in self.delta:
                if u == x and w == EPS:
                    t = t | E[v]
            if t != E[x]:
                E[x] = t
                for u, w, v in self.delta:
                    if v == x and w == EPS and u not in worklist:
                        worklist.append(u)
            iter_cnt += 1
        print(E, iter_cnt)

    def subset_construction(self):
        """子集构造法"""

        def get_delta(q: list, c: str) -> list:
            S = []
            for x in q:
                for u, w, v in self.delta:
                    if u == x and w == c and v not in S:
                        S.append(v)
            return S

        def eps_closure(S: list) -> list:
            # 不动点计算
            worklist = S.copy()
            S = S.copy()
            while worklist:
                x = worklist.pop(0)
                for u, w, v in self.delta:
                    if u == x and w == EPS and v not in S:
                        S.append(v)
                        worklist.append(v)
            return sorted(S)

        Q = [eps_closure([self.s0])]
        worklist = [Q[0]]
        T = {}
        TT = {}

        while worklist:
            q = worklist.pop(0)
            T[Q.index(q)] = {}
            TT[Q.index(q)] = {}
            for c in self.sigma:
                t = eps_closure(get_delta(q, c))
                T[Q.index(q)][c] = t if t not in Q else f"{Q.index(t)}"
                if t not in Q:
                    Q.append(t)
                    worklist.append(t)
                TT[Q.index(q)][c] = Q.index(t)
        delta = [(str(u), str(w), str(v)) for u, items in TT.items() for w, v in items.items()]
        dfa = DFA(
            S=[str(i) for i in T.keys()],
            sigma=self.sigma,
            delta=delta,
            s0=str(0),
            A=[str(i) for i in range(len(Q)) if set(self.A) & set(Q[i])],
        )
        return dfa, Q, T


class DFA(NFA):

    def move(self, s: str, w: str):
        return self.G[s][w][0] if w in self.G[s] else None

    def hopcroft(self):
        def split_into(s, c):
            """将状态集s按照c分割成s1和s2"""
            dst = {}
            s1, s2 = set(), set()
            for u in s:
                v = self.move(u, c)
                if v:
                    # 经c可转移
                    dst[u] = v
                    s1.add(u)
                else:
                    # 经c不可转移
                    s2.add(u)
            if not s2:
                # 不存在经c不可转移的点,按照转移点所属集合切分
                groupby = {}
                for k, v in dst.items():
                    groupby.setdefault(stateMap[v], []).append(k)
                if len(groupby) > 1:
                    print(s, c, list(groupby.values()))
                s1 = set(next(iter(groupby.values())))
                s2 = s - s1
            return frozenset(s1), frozenset(s2)

        def split(s):
            for c in self.sigma:
                s1, s2 = split_into(s, c)
                if s2:
                    return {s1, s2}
            return {s}

        T = {frozenset(self.A), frozenset(self.S) - frozenset(self.A)}
        P = set()
        stateMap = {}
        while T != P:
            stateMap = {x: str(i) for i, s in enumerate(T) for x in s}
            P = T
            T = set()
            for p in P:
                T = T | split(p)
        T = list(map(list, T))
        print(T)
        min_dfa = DFA(
            S=[str(i) for i in range(len(T))],
            sigma=self.sigma,
            delta=list({(stateMap[u], c, stateMap[v]) for u, c, v in self.delta}),
            s0=stateMap[self.s0],
            A=list({stateMap[i] for i in self.S if i in self.A}),
        )
        return min_dfa

    def hopcroft1(self):
        partition = [set(self.A), set(self.S) - set(self.A)]
        worklist = [set(self.A), set(self.S) - set(self.A)]
        PI = []
        while worklist:
            s = worklist.pop()
            for c in self.sigma:
                image = {u for u, w, v in self.delta if w == c and v in s}
                for q in partition:
                    q1 = q & image
                    q2 = q - q1
                    if q1 and q2:
                        PI.append((partition, q, c, q1, q2))
                        partition.remove(q)
                        partition.append(q1)
                        partition.append(q2)
                        if q in worklist:
                            worklist.remove(q)
                            worklist.append(q1)
                            worklist.append(q2)
                        else:
                            worklist.append(q1 if len(q1) <= len(q2) else q2)
        PI.append((partition, {}, None, {}, {}))
        stateMap = {x: min(s) for s in partition for x in s}
        min_dfa = DFA(
            S=[str(i) for i in stateMap.values()],
            sigma=self.sigma,
            delta=list({(stateMap[u], c, stateMap[v]) for u, c, v in self.delta}),
            s0=str(stateMap[self.s0]),
            A=list({str(stateMap[i]) for i in self.A}),
        )
        return min_dfa, PI


if __name__ == "__main__":
    nfa = NFA.load("./input/FA/test1_FA.txt")
    nfa.dot("nfa", view=True)
    dfa, Q, T = nfa.subset_construction()
    print("----------子集构造法----------")
    df1 = pd.DataFrame(T, index=dfa.sigma).T.rename_axis("集合名称")
    df1.insert(0, "DFA状态子集", Q)
    print(df1)
    dfa.dot("dfa", view=True)
    print("----------Hopcroft算法----------")
    min_dfa, PI = dfa.hopcroft1()
    print(pd.DataFrame(PI))
    min_dfa.dot("min_dfa", view=True)

File: utils/FA.js
import fs, { symlink } from "fs";
import { digraph } from "graphviz";
import path from "path";

const EPS = "ε";
const EOF = "$";

/**
 * 获取A中是否有x的索引
 * @param {Array<Array>} A
 * @param {Array} x
 * @returns
 */
const indexOf = (A, x) => {
    return A.findIndex((a) => JSON.stringify([...a]) === JSON.stringify([...x]));
};

class FA {
    /**
     * @param {Array<string>} S
     * @param {Array<Array<string>>} sigma
     * @param {Array} delta
     * @param {string} s0
     * @param {Array<string>} A
     */
    constructor(S, sigma, delta, s0, A) {
        this.S = S;
        this.sigma = sigma;
        this.delta = delta;
        this.s0 = s0;
        this.A = A;

        // 构建状态转移图 G
        this.G = {};
        for (let state of S) {
            this.G[state] = {};
        }
        for (let [u, w, v] of delta) {
            if (!this.G[u][w]) this.G[u][w] = [];
            this.G[u][w].push(v);
        }
    }

    /**
     * 加载FA
     * @param {string} filename 文件名
     * @param {FA|NFA|DFA} cls
     * @returns {FA} FA实例
     */
    static load(filename, cls = FA) {
        let content = fs.readFileSync(filename, "utf8").replace(/eps/g, EPS);
        const lines = content
            .trim()
            .split("\n")
            .map((line) => line.trim())
            .filter((line) => line);
        const S = lines[0].split(" ");
        const sigma = lines[1].split(" ");
        const delta = lines.slice(2, -2).map((item) => item.split(" "));
        const s0 = lines[lines.length - 2];
        const A = lines[lines.length - 1].split(" ");
        return new cls(S, sigma, delta, s0, A);
    }

    move(s, w) {
        return this.G[s]?.[w] || [];
    }

    /**
     *
     * @param {string} filename - 文件名
     * @returns {string} - dot源码
     */
    dot(filename = "fa") {
        let dot = `digraph ${filename} {\n`;
        dot += "    rankdir=LR;\n";
        this.S.forEach((state) => {
            state = state.replace('"', '\\"');
            const isStart = state === this.s0;
            const isAccept = this.A.includes(state);
            let style = [];
            style.push(isAccept ? "shape=doublecircle" : "shape=circle");
            dot += `    ${state} [${style.join(",")}];\n`;
        });
        // 添加转移
        let delta = {};
        this.delta.forEach(([from, symbol, to]) => {
            from = from.replace('"', '\\"');
            to = to.replace('"', '\\"');
            symbol = symbol.replace('"', '\\"');
            delta[from] ??= {};
            delta[from][to] ??= [];
            delta[from][to].push(symbol);
        });
        dot += Object.entries(delta).flatMap(([from, item]) =>
            Object.entries(item).map(([to, symlbols]) => `    ${from} -> ${to} [label="${symlbols.join(',')}"];`)
        ).join('\n');

        dot += "}";
        return dot;
        // const dot = new digraph(filename);
        // for (let u of this.S) {
        //     dot.addNode(u).set("label", u);
        //     dot.getNode(u).set("shape", "circle");
        // }
        // for (let a of this.A)
        //     dot.getNode(a).set("shape", "doublecircle");
        // for (let [u, w, v] of this.delta)
        //     dot.addEdge(u, v).set("label", w);
        // fs.writeFileSync(`./output/${filename}.dot`, dot.to_dot());
        // return dot.to_dot();
    }
}

class NFA extends FA {
    /**
     * 加载NFA
     * @param {string} filename 文件名
     * @returns {NFA} NFA实例
     */
    static load(filename) {
        return super.load(filename, NFA);
    }

    /**
     * 子集构造法
     * @returns {[DFA, Array<Array<string>>, Object]} - [DFA, DFA状态子集列表, DFA状态转移表]
     */
    subsetConstruction() {
        /**
         * 获取经c状态转移
         * @param {Array<string>} q
         * @param {string} c
         * @returns {Set<string>}
         */
        const getDelta = (q, c) => {
            let S = new Set();
            for (let state of q) {
                for (let [u, w, v] of this.delta) {
                    if (u === state && w === c) {
                        S.add(v);
                    }
                }
            }
            return S;
        };

        /**
         * 求eps闭包
         * @param {Set<string>} S
         * @returns {Array<string>}
         */
        const eps_closure = (S) => {
            let worklist = [...S];
            let closureSet = [...S];

            while (worklist.length > 0) {
                let x = worklist.shift();
                for (let [u, w, v] of this.delta) {
                    if (u === x && w === EPS && !closureSet.includes(v)) {
                        closureSet.push(v);
                        worklist.push(v);
                    }
                }
            }
            return closureSet.sort();
        };

        let Q = [eps_closure([this.s0])];
        let worklist = [Q[0]];
        let T = {};
        let TT = {};

        while (worklist.length > 0) {
            const q = worklist.shift();
            T[indexOf(Q, q)] = {};
            TT[indexOf(Q, q)] = {};
            for (const c of this.sigma) {
                const t = eps_closure(getDelta(q, c));
                if (t.length === 0) continue;
                T[indexOf(Q, q)][c] = indexOf(Q, t) === -1 ? t : indexOf(Q, t);
                if (indexOf(Q, t) === -1) {
                    Q.push(t);
                    worklist.push(t);
                }
                TT[indexOf(Q, q)][c] = indexOf(Q, t)
            }
        }

        let delta = Object.entries(TT).flatMap(([s, items]) =>
            Object.entries(items).map(([c, t]) => [s, c, String(t)])
        );

        const A = Q.map((_, i) => i)
            .filter((i) => Q[i].some((state) => this.A.includes(state)))
            .map((i) => String(i));

        return [
            new DFA(
                Q.map((_, i) => String(i)),
                this.sigma,
                delta,
                String(0),
                A
            ),
            Q,
            T,
        ];
    }
}

class DFA extends NFA {
    /**
     * 加载DFA
     * @param {string} filename 文件名
     * @returns {DFA} DFA实例
     */
    static load(filename) {
        return super.load(filename, DFA);
    }
    move(s, w) {
        return this.G[s]?.[w]?.[0] || null;
    }

    /**
     * hopcroft算法
     * @returns {[DFA, Array<Array<string>>]} - [最小化DFA, DFA构建过程]
     */
    hopcroft() {
        let partition = [
            new Set(this.A),
            new Set(this.S.filter((s) => !this.A.includes(s))),
        ];
        let worklist = [new Set(this.A)];
        let PI = [];
        while (worklist.length > 0) {
            let s = worklist.pop();
            for (let c of this.sigma) {
                let image = new Set();
                for (let [u, w, v] of this.delta) {
                    if (w === c && s.has(v)) {
                        image.add(u);
                    }
                }
                for (let q of partition) {
                    let q1 = new Set([...q].filter((x) => image.has(x)));
                    let q2 = new Set([...q].filter((x) => !image.has(x)));

                    if (q1.size > 0 && q2.size > 0) {
                        PI.push([
                            partition.map((item) => Array.from(item)),
                            Array.from(q),
                            c,
                            Array.from(q1),
                            Array.from(q2),
                        ]);
                        partition.splice(indexOf(partition, q), 1, q1, q2);
                        if (indexOf(worklist, q) != -1) {
                            worklist.splice(indexOf(worklist, q), 1, q1, q2);
                        } else {
                            worklist.push(q1.size <= q2.size ? q1 : q2);
                        }
                    }
                }
            }
        }
        PI.push([partition.map((item) => Array.from(item)), [], null, [], []]);
        let stateMap = {};
        partition.forEach((group, i) => {
            const x = Math.min(...Array.from(group).map((item) => Number(item)));
            for (let s of group) {
                stateMap[s] = String(x);
            }
        });
        let S = Object.values(stateMap);
        // let S = partition.map((_, i) => String(i));
        let delta = [];
        for (let [u, w, v] of this.delta) {
            if (indexOf(delta, [stateMap[u], w, stateMap[v]]) === -1) {
                delta.push([stateMap[u], w, stateMap[v]]);
            }
        }
        let A = [...new Set(this.A.map((s) => stateMap[s]))];
        return [new DFA(S, this.sigma, delta, stateMap[this.s0], A), PI];
    }
}
export { FA, NFA, DFA };

// const nfa = NFA.load(path.join('input', 'FA', 'test1_FA.txt'));
// console.log("NFA:")
// console.table(nfa.G);
// nfa.dot("nfa");

// const [dfa, Q, T] = nfa.subsetConstruction();
// console.log("subsetConstruction:");
// console.table(T);
// console.log("DFA:")
// console.table(dfa.G); // 输出DFA的状态、字母表、转移函数、初始状态和接受状态
// dfa.dot("dfa");

// const [minDfa, PI] = dfa.hopcroft();
// console.log("hopcroft:")
// console.table(PI);
// console.log("minDfa:")
// console.table(minDfa.G); // 输出最小化DFA的状态、字母表、转移函数、初始状态和接受状态
// minDfa.dot("min_dfa");
